### PR TITLE
bpo-46055: Speed up binary shifting operators

### PR DESCRIPTION
**Pull request by xuxinhang**
Inspired by https://bugs.python.org/issue44946, I found there were no special shortcuts for shifting operation applied to "medium value" long object. So I modified CPython's VM to accelerate my python project where there is plenty of binary shifting operation. I guess somebody else might also need it.

Thanks.

https://bugs.python.org/issue46055

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@xuxinhang

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by sweeneyde** [comment]
Hi @xuxinhang, it would be great if you could open an issue on bugs.python.org to associate with this PR, and change the PR title accordingly.

Also, please make sure to get the CLA signed. Thanks!

**Comment by xuxinhang** [comment]
Sorry, I have create a new issue in the bug tracker and I'm waiting for CLA getting verified. Besides, the building in Ubuntu  doesn't complain segment fault now.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: Why is this being changed? It seems to be adding branches into a very hot code path.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: It would make more sense to me to just call _PyLong_New(Py_ABS(...)) where applicable rather than changing this function. Other call sites already ensure size >= 0.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: From https://en.cppreference.com/w/cpp/language/operator_arithmetic :

"""For negative a, the value of a >> b is implementation-defined (in most implementations, this performs arithmetic right shift, so that the result remains negative)."""

I think we want to avoid undefined behavior, even if the machines with the bad behavior are rare.

You can probably do something like `-(Py_SIZE(a) < 0)`, and compilers will likely optimize that anyway.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: Isn't this incorrect?

`(-8) >> 4` should be -1, but `-(8 >> 4)` is 0.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: [PEP 7](https://www.python.org/dev/peps/pep-0007/) says all newly-written code should include `{braces}` around blocks.

**Review by sweeneyde — COMMENTED** [review]
I didn't do a complete review, but here are a few comments.

It would also be nice to see some benchmarks for the change.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: The question of undefined behavior applies here as well.

Comment (sweeneyde) [reply Objects/longobject.c]: You can probably use [Py_ARITHMETIC_RIGHT_SHIFT](https://github.com/python/cpython/blob/cb8f491f46e262549f6c447b31625cab7c20a60a/Include/pyport.h#L300)

**Comment by xuxinhang** [comment]
### BENCHMARKS
I use `timeit` to measure time costs and any other operators or calls are excluded. For each testcase, the former is `dcd2796` and the latter is this PR's base `036bbb1`. 

64-bit Release building. Run in Windows 10 1709 (64-bit)

```
python -m timeit " i = 1; i <<= 3; i >>= 3"  # small value (cost down by 36%)
5000000 loops, best of 5: 92.7 nsec per loop
2000000 loops, best of 5: 145 nsec per loop

python -m timeit " i = 1; i <<= 10; i >>= 10"  # medium value (-25%)
2000000 loops, best of 5: 114 nsec per loop
2000000 loops, best of 5: 151 nsec per loop

python -m timeit " i = 1; i <<= 100; i >>= 100"  # big value  (-12%)
1000000 loops, best of 5: 209 nsec per loop
1000000 loops, best of 5: 238 nsec per loop
```

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: That's a wrong language expression. It's mine.

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: I have thought this could avoid `MUL` instruction, but the benchmark proves this change brings no better or worse performance. I'll drop it.

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: I used `_PyLong_New(Py_ABS(...))` in commit `403652a`. But then I felt it weird that `_PyLong_New` can't accept a negative value as size. If we regard `size` as the allocated memory size, it shouldn't be assigned to `ob_size`. If we regard `size` as the value of `ob_size`, it should be allowed to be negative. That seems a little confusing, in my view.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: Isn't this is undefined behavior as well?

Comment (xuxinhang) [reply Objects/longobject.c]: [PyLong_MASK](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Include/cpython/longintrepr.h#L64) is an unsigned value (defined with type `digit`) as the mask of `PyLong->ob_digit`. Maybe it can be regarded as a special `ob_digit` value somewhat.

Comment (sweeneyde) [reply Objects/longobject.c]: I mean if medium_value(a) is negative

Comment (xuxinhang) [reply Objects/longobject.c]: Thanks. Fixed.

File: Objects/longobject.c
Comment (sweeneyde) [reply Objects/longobject.c]: From what I understand, on most modern CPUs, a multiplication isn't *that* much more expensive than (say) addition, but branches like this can be quite expensive if they are unpredictable.

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c]: There's a bug here: `(-10000000000000000000) >> 100000` should return `-1`, not `0`. We should also add a test case to Lib\test\test_long.py for this and related cases.

Comment (sweeneyde) [reply Objects/longobject.c]: This could be `return get_small_int(-(Py_SIZE(a) < 0));`

File: Objects/longobject.c
Comment (sweeneyde) [inline Objects/longobject.c:4584]: I would move this comment one line up. PEP 7 says "Comments go before the code they describe."

**Review by mdickinson — COMMENTED** [review]
There are a few changes in this PR that aren't directly related to the stated goal of speeding up the shift operators. Please could you revert the unrelated changes, so that we can focus on the shift operations? That would make the PR easier to review and evaluate.

File: Objects/longobject.c
Comment (mdickinson) [reply Objects/longobject.c]: Indeed, the multiplication spelling was deliberately introduced recently in a PR targeting speed of integer operations. See #27832. Please revert this change.

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: Please revert; this is unrelated to the main purpose of the PR.

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: Let's revert this one too. We can re-introduce it in a separate PR, if performance measurements show that it's worth it.

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: This may not be big enough in the negative case. Consider the case of `-(2**60 - 1) >> 30` (assuming 30-bit digits). Then `Py_SIZE(a)` is `2`, `wordshift` is `1`, but we need two digits for the result.

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: I think this whole block could be simpler. If `m = medium_value(a)`, then if `wordshift == 0`, we want `m >> remshift` if `m` is nonnegative, and `~(~m >> remshift)` if `m` is negative. Then there's no need for `Py_ARITHMETIC_RIGHT_SHIFT`, since we're only ever shifting non-negative values.  To convert the result back to a `PyLong`, I'd suggest using `_PyLong_FromSTwoDigits`.

Comment (mdickinson) [reply Objects/longobject.c]: To be clear, I'd suggest something along the lines of:
```c
    if (IS_MEDIUM_VALUE(a)) {
        stwodigits sval;
        stwodigits x = medium_value(a);
        if (wordshift == 0) {
            sval = x < 0 ? ~(~x >> remshift) : x >> remshift;
        }
        else {
            sval = x < 0 ? -1 : 0;
        }
        return _PyLong_FromSTwoDigits(sval);
    }
```

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: Why is the `(a->ob_digit[0] & ~(PyLong_MASK >> remshift)) == 0` check necessary?

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: I'd suggest splitting into negative and nonnegative cases. It'll be easier to read that way, and a good compiler can probably figure out how to recombine them if that's more efficient for the processor. (And the code as written isn't strictly portable. There are some crazy corner cases in C's integer handling: the left operand of `<<` here doesn't even necessarily have unsigned type, because if `twodigits` is smaller than `int` (which is certainly allowed by the standard) then the left operand gets promoted to an `int`. We're unlikely to run into this case in practice, but we might as well go for the direct and obviously correct spelling anyway.)

**Review by mdickinson — CHANGES_REQUESTED** [review]
Hi @xuxinhang. Thanks for the contribution - it seems a good idea in principle to speed up the small left shift and right shift cases.

I'd probably avoid the reworking of the non-small cases for negative numbers, for the moment, or move that into a separate PR.

There is quite a lot of cleanup needed before this PR can be ready - please see the various comments from me and from @sweeneyde.

File: Objects/longobject.c
Comment (sweeneyde) [reply Objects/longobject.c]: Compiler Explorer is showing that gcc and clang eliminate that `~` branch on x64, but MSVC does not.

https://godbolt.org/z/8hPvGW3WT

File: Objects/longobject.c
Comment (sweeneyde) [reply Objects/longobject.c]: I think it's necessary because `(2**20) << 20` won't fit in a single 30/32-bit digit, and this checks for cases like that.

Comment (sweeneyde) [reply Objects/longobject.c]: Oh I guess it's not necessary if we use `_PyLong_FromSTwoDigits`

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: The checking `a->ob_digit[0] & ~(PyLong_MASK >> remshift) == 0` ensures no bit overflowing or discarding, that is valid bits would stay in the range covered by twodigit.

Besides, I'm sorry but I couldn't catch the idea about "splitting into negative and nonnegative cases" - it's both C's left shifting that applied to "medium value". Could you please explain it further?

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: `z = _PyLong_New(newsize)` is a helper long object, whose value is assigned as `(2**60 - 1) >> 30` instead of `-(2**60 - 1) >> 30`. The `long_sub()` below gives the final result.

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: It's necessary to ensure the shifting result fits `twodigits` or "medium value" before using `_PyLong_FromSTwoDigits` or something else.

Comment (mdickinson) [reply Objects/longobject.c]: `medium_value(a) << remshift` will always fit in an `stwodigits`, though. I think it's safe to drop this condition.

File: Objects/longobject.c
Comment (mdickinson) [reply Objects/longobject.c]: So the problem is that it takes some mental effort to understand why casting to `twodigits`, shifting, and then casting back to `stwodigits` gives the correct result in the negative case. Moreover, there are corner cases (admittedly unlikely to occur in practice, but still theoretically possible) where it's still undefined behaviour, if the integer promotions promote an `twodigits` to `int`. See C99 §6.3.1.1, paragraph 2.

Because of those two things, I'd prefer a more direct spelling: if `medium_value(a)` is nonnegative, do `medium_value(a) << remshift` (without any casting). If `medium_value(a)` is negative, do `-((-medium_value(a)) << remshift)` (again without any casting). So something like `stwodigits sval = m < 0 ? -(-m << remshift) : m << remshift`, where `m` is `medium_value(a)`.

File: Objects/longobject.c
Comment (mdickinson) [reply Objects/longobject.c]: Ah, okay; thanks.

I'd propose dropping this set of changes from the PR and focusing on the small int case; we can re-introduce these changes later in a separate PR. I think they'll require their own round of iterations, and I'd rather get the small int improvements in quickly.

File: Objects/longobject.c
Comment (mdickinson) [reply Objects/longobject.c]: @sweeneyde: Nice; thanks! I think I'm happy with that - there's a good chance some future version of MSVC will catch up.

Comment (mdickinson) [reply Objects/longobject.c]: @sweeneyde: Looks like this spelling fares better on MSVC:
```c
    int shift = wordshift == 0 ? remshift : PyLong_SHIFT;
    stwodigits sval = x < 0 ? ~(~x >> shift) : x >> shift;
```



https://godbolt.org/z/chfKj9E9W

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: This comment is becoming increasingly inaccurate and confusing (exactly *which* part of the code does it apply to?), and curiously specific (since so much else in this module is also due to Tim Peters). @tim-one: would you object to it being dropped?

File: Objects/longobject.c
Comment (mdickinson) [inline Objects/longobject.c]: Suggestion that may remove a branch on common compilers:

```suggestion
        int shift = wordshift == 0 ? remshift : PyLong_SHIFT;
        x = Py_ARITHMETIC_RIGHT_SHIFT(stwodigits, medium_value(a), shift);
```

File: Objects/longobject.c
Comment (tim-one) [reply Objects/longobject.c]: Get rid of it - looks like @serhiy-storchaka added it in 2019, presumably to make it clearer at the time which parts they were adding.

File: Objects/longobject.c
Comment (xuxinhang) [reply Objects/longobject.c]: Ah, Okay. Thanks.

File: Objects/longobject.c
Comment (serhiy-storchaka) [reply Objects/longobject.c]: I did not add it, I just moved some code before and after this line, and git was confused.

It was added by GvR in 1997 (f2e499b1d760076f519b618105c112495d6f7911).

Comment (tim-one) [reply Objects/longobject.c]: OK! After a quarter century, I expect it's served its purpose & can be retired now 😄

**Review by mdickinson — CHANGES_REQUESTED** [review]
@xuxinhang This is looking good - thank you! A couple more things:

- Please could you revert the unrelated spacing changes? Just as a general principle, it's usually a good idea to keep a PR minimal and focused on the necessary changes - that makes it easier to understand and review, and avoids spurious merge conflicts and history confusion.
- Please could you add a news entry? (See [instructions for using blurb](https://devguide.python.org/committing/?highlight=blurb#updating-news-and-what-s-new-in-python) in the developer guide.)
- Feel free to add your own name to the `Misc/ACKS` file if you want to. (This is completely optional.)

File: Objects/longobject.c
Comment (mdickinson) [reply Objects/longobject.c]: @xuxinhang This change still needs to be reverted - it's unrelated to the main purpose of the PR.

**Comment by xuxinhang** [comment]
Thanks. 
——
I have made the requested changes; please review again

**Comment by mdickinson** [comment]
@xuxinhang: Almost there. Please could you undo the changes to the `medium_value` function?

**Review by sweeneyde — APPROVED** [review]

File: Misc/NEWS.d/next/Core and Builtins/2021-12-24-20-21-45.bpo-46055.R0QMVQ.rst
Comment (mdickinson) [inline Misc/NEWS.d/next/Core and Builtins/2021-12-24-20-21-45.bpo-46055.R0QMVQ.rst]:
```suggestion
Speed up shifting operation involving integers less than
:c:macro:`PyLong_BASE`. Patch by Xinhang Xu.
```

**Review by mdickinson — APPROVED** [review]
LGTM! Merging when the CI completes. Thank you!

**Comment by mdickinson** [comment]
Some sample timings on my machine. "main" is at commit b8de8b7039cee47465b2af9950b0b9ed2d3f2903. "here" is this branch, after a merge with main. Timings performed on an Intel MacBook Pro, macOS 10.14.6, with `./python.exe -m timeit -s "i, s = 1, 3" "i << s"` or `./python.exe -m timeit -s "i, s = 1, 3" "i >> s"`, with suitable values of `i` and `s` in the setup.

For `1<<3`:
main: 5000000 loops, best of 5: 38.2 nsec per loop
here: 10000000 loops, best of 5: 24.7 nsec per loop
55% faster

For `1234567 >> 13`:
main: 5000000 loops, best of 5: 38.2 nsec per loop
here: 10000000 loops, best of 5: 24.7 nsec per loop
55% faster

(No, those numbers aren't copy and paste errors - I really did get the exact same timings.)

For `1234567 >> 123`:
main: 10000000 loops, best of 5: 23.3 nsec per loop
here: 10000000 loops, best of 5: 25.3 nsec per loop
8% slower (though this seems to be within random variation)

For `-1234567 >> 13`:
main: 5000000 loops, best of 5: 67.5 nsec per loop
here: 10000000 loops, best of 5: 36.2 nsec per loop
86% faster

**Comment by mdickinson** [comment]
The `Tests / Windows (x64) (pull_request)` check seems to be stuck. Closing and re-opening to re-trigger the build.

**Comment by serhiy-storchaka** [comment]
I suggest to merge #30243 first and compare this PR with a new base.

**Comment by mdickinson** [comment]
Some new sample timings, now that #30243 has been merged into master, using the same machine and methodology as above. It's interesting that `1234567 >> 13` gets a significant speedup on my machine, but `12345678 >> 13` doesn't; looks like in the second case, the time is dominated by the time to construct a new PyLong for the result. (In the first case, the result is small.)

Comparing commits:
here: 59672cfbb4c013a602676df0157be377174deb45
main: 360fedc2d2ce6ccb0dab554ef45fe83f7aea1862

`1 << 3`: ~43% faster
- here: 10000000 loops, best of 5: 26.7 nsec per loop
- main: 10000000 loops, best of 5: 38.1 nsec per loop

`1234567 >> 13`: ~36% faster
- here: 10000000 loops, best of 5: 26.6 nsec per loop
- main: 10000000 loops, best of 5: 36.2 nsec per loop

`12345678 >> 13`: ~6% faster
- here: 10000000 loops, best of 5: 34.3 nsec per loop
- main: 10000000 loops, best of 5: 36.4 nsec per loop

`-1234567 >> 13`: ~77% faster
- here: 5000000 loops, best of 5: 38.8 nsec per loop
- main: 5000000 loops, best of 5: 68.6 nsec per loop

**Pull request merged**
